### PR TITLE
Adding pytorch unbatched code for 03-emb

### PR DESCRIPTION
**Pull request by MysteryVaibhav**

### CODE REVIEW

**Comment by danishpruthi** [comment]
Thanks Vaibhav, I will have a look once.

**Comment by MysteryVaibhav** [comment]
@danishpruthi Thanks. Since, the overall time for running the code on CPU was taking forever, I ran both the pytorch version and the dynet version for 5 iterations on a subset of data for basic validation. Seems like the pytorch version is _4x_ slower for cbow and _15x_ slower for skip. Loss wise the numbers are consistent. Also, there might be a bug in the dynet version of skip-gram model. 

**cbow-py**
```
started iter 0
iter 0: train loss/word=7.6407, ppl=2081.2178, time=21.92s
iter 0: dev loss/word=7.9544, ppl=2848.0582, time=0.57s
saving embedding files
started iter 1
iter 1: train loss/word=5.1192, ppl=167.1964, time=22.81s
iter 1: dev loss/word=8.0769, ppl=3219.1971, time=0.63s
saving embedding files
started iter 2
iter 2: train loss/word=3.3664, ppl=28.9742, time=22.94s
iter 2: dev loss/word=8.3735, ppl=4330.5775, time=0.58s
saving embedding files
started iter 3
iter 3: train loss/word=1.9536, ppl=7.0541, time=24.51s
iter 3: dev loss/word=9.0460, ppl=8484.3998, time=0.61s
saving embedding files
started iter 4
iter 4: train loss/word=1.2298, ppl=3.4206, time=24.16s
iter 4: dev loss/word=9.4130, ppl=12246.8663, time=0.59s
 ```

**cbow-dy**
```
iter 0: train loss/word=8.5192, ppl=5009.9690, time=5.55s
iter 0: dev loss/word=8.3132, ppl=4077.2017, time=1.27s
saving embedding files
started iter 1
iter 1: train loss/word=7.3918, ppl=1622.6030, time=5.17s
iter 1: dev loss/word=8.1308, ppl=3397.4341, time=1.18s
saving embedding files
started iter 2
iter 2: train loss/word=6.8121, ppl=908.7425, time=5.18s
iter 2: dev loss/word=8.0865, ppl=3250.3900, time=1.15s
saving embedding files
started iter 3
iter 3: train loss/word=6.4230, ppl=615.8689, time=5.29s
iter 3: dev loss/word=8.0603, ppl=3166.1412, time=1.17s
saving embedding files
started iter 4
iter 4: train loss/word=6.1320, ppl=460.3542, time=5.28s
iter 4: dev loss/word=8.0951, ppl=3278.2668, time=1.46s
 ```

**skip-py**
```
started iter 0
iter 0: train loss/word=2.2175, ppl=9.1844, time=71.35s
iter 0: dev loss/word=2.3332, ppl=10.3108, time=0.36s
saving embedding files
started iter 1
iter 1: train loss/word=1.1389, ppl=3.1232, time=75.66s
iter 1: dev loss/word=2.1954, ppl=8.9840, time=0.29s
saving embedding files
started iter 2
iter 2: train loss/word=0.6602, ppl=1.9352, time=74.14s
iter 2: dev loss/word=2.1366, ppl=8.4709, time=0.31s
saving embedding files
started iter 3
iter 3: train loss/word=0.4383, ppl=1.5500, time=72.24s
iter 3: dev loss/word=2.1052, ppl=8.2089, time=0.30s
saving embedding files
started iter 4
iter 4: train loss/word=0.3148, ppl=1.3699, time=72.73s
iter 4: dev loss/word=2.0865, ppl=8.0568, time=0.29s
 ```

**skip-dy**
```
iter 0: train loss/word=-39.3036, ppl=0.0000, time=5.78s
iter 0: dev loss/word=-42.2779, ppl=0.0000, time=1.37s
saving embedding files
started iter 1
iter 1: train loss/word=-58.8366, ppl=0.0000, time=5.25s
iter 1: dev loss/word=-62.7666, ppl=0.0000, time=1.17s
saving embedding files
started iter 2
iter 2: train loss/word=-104.0526, ppl=0.0000, time=5.23s
iter 2: dev loss/word=-98.7395, ppl=0.0000, time=1.19s
saving embedding files
started iter 3
iter 3: train loss/word=-176.7759, ppl=0.0000, time=5.26s
iter 3: dev loss/word=-150.2649, ppl=0.0000, time=1.22s
saving embedding files
started iter 4
iter 4: train loss/word=-510.4765, ppl=0.0000, time=6.45s
iter 4: dev loss/word=-794.8438, ppl=0.0000, time=1.23s
 ```

**Comment by danishpruthi** [comment]
@MysteryVaibhav: I had a look at the skipgram model, I think (I might be wrong) you are making a mistake in when you are calculating `logsigmoid`. The goal here is to make the current word predict its context, and when we need to make sure that the _probability_ (or log of probability) of these context words are high. But, in your implementation, you are optimizing the unnormalized score of the context without looking at other words in the vocabulary. 

https://github.com/neubig/nn4nlp-code/blob/6ae371f9cdf2dc939ed8da1bd5aa8ca8457206c7/03-wordemb-pytorch/wordemb-skip.py#L22-L28

Specifically, In line 27, I think you are only taking a log-sigmoid over the target index rather than the whole vocabulary. 

Does this make sense?

File: 03-wordemb-pytorch/wordemb-skip.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+            for direction in [-1, 1]:
+                context_id = sent[i + direction * j] if 0 <= i + direction * j < len(sent) else S
+                context = torch.tensor([context_id]).type(type)   # Tensor for context word
+                loss = model(c, context)
+                if not inference:
+                    # Back prop while training only
+                    optimizer.zero_grad()
+                    loss.backward()
+                    optimizer.step()
+                total_loss += loss.data.cpu().item()
```
Comment (danishpruthi) [inline 03-wordemb-pytorch/wordemb-skip.py]: A nit pick: but for single valued tensors, you can simply do `loss.item()`, you don't need to transfer the data to cpu.

File: 03-wordemb-pytorch/wordemb-skip.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+class Skip(torch.nn.Module):
+    def __init__(self, nwords, emb_size):
+        super(Skip, self).__init__()
+
+        """ word embeddings """
+        self.word_embedding = torch.nn.Embedding(nwords, emb_size)
+        """ context embeddings"""
+        self.context_embedding = torch.nn.Embedding(nwords, emb_size)
+
+    def forward(self, word_pos, context_pos):
```
Comment (danishpruthi) [inline 03-wordemb-pytorch/wordemb-skip.py]: @MysteryVaibhav: I had a look at the skipgram model, I think (I might be wrong) you are making a mistake in when you are calculating `logsigmoid`. The goal here is to make the current word predict its context, and when we need to make sure that the _probability_ (or log of probability) of these context words are high. But, in your implementation, you are optimizing the unnormalized score of the context without looking at other words in the vocabulary. 

https://github.com/neubig/nn4nlp-code/blob/6ae371f9cdf2dc939ed8da1bd5aa8ca8457206c7/03-wordemb-pytorch/wordemb-skip.py#L22-L28

Specifically, In line 27, I think you are only taking a log-sigmoid over the target index rather than the whole vocabulary. 

Does this make sense?

File: 03-wordemb-pytorch/wordemb-cbow.py [hunk 3]
```diff
@@ -0,0 +1,122 @@
+from collections import defaultdict
+import math
+import time
+import random
+import torch
+
+
+class CBoW(torch.nn.Module):
```
Comment (danishpruthi) [inline 03-wordemb-pytorch/wordemb-cbow.py]: Maybe the change the name to `WordEmbCbow` to make it more specific.

File: 03-wordemb-pytorch/wordemb-cbow.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+    # Step through the sentence
+    total_loss = 0
+    for i in range(N, len(sent) + N):
+        # c is the context vector
+        c = torch.tensor(padded_sent[i - N:i] + padded_sent[i + 1:i + N + 1]).type(type)
+        t = torch.tensor([padded_sent[i]]).type(type) # This is the target vector
+        log_prob = model(c)
+        loss = criterion(log_prob, t)   # loss for predicting target from context vector
+        if not inference:
```
Comment (danishpruthi) [inline 03-wordemb-pytorch/wordemb-cbow.py]: in the Dynet example, the backprop happens after summing all the losses for a given sentence rather than after every word seen, I think it would be best to make it consistent with that.

Comment (danishpruthi) [reply 03-wordemb-pytorch/wordemb-cbow.py]: Also, I think the choice of `inference` is confusing. Technically, you are doing inference during training and dev both. Maybe just change `inference` flag with the opposite `need_to_train` to avoid the confusion.

**Review by danishpruthi — CHANGES_REQUESTED** [review]
@MysteryVaibhav: I think this is mostly in good shape (except 1 major comment), but please have a look at the suggestions. Once you have fixed those, please reply to the thread, and then I can approve it.

File: 03-wordemb-pytorch/wordemb-skip.py (hunk 1 above)
Comment (MysteryVaibhav) [reply 03-wordemb-pytorch/wordemb-skip.py]: Changed to .item()

File: 03-wordemb-pytorch/wordemb-skip.py (hunk 2 above)
Comment (MysteryVaibhav) [reply 03-wordemb-pytorch/wordemb-skip.py]: @danishpruthi : As discussed with you earlier, this was just the word2vec impl. of minimizing the cosine distance b/w a word and its context word (without the negative sampling part). I have changed it to match the exact dynet implementation which involves taking a softmax over the context words.

File: 03-wordemb-pytorch/wordemb-cbow.py (hunk 3 above)
Comment (MysteryVaibhav) [reply 03-wordemb-pytorch/wordemb-cbow.py]: Done

File: 03-wordemb-pytorch/wordemb-cbow.py (hunk 4 above)
Comment (MysteryVaibhav) [reply 03-wordemb-pytorch/wordemb-cbow.py]: @danishpruthi : Thanks for pointing this out. I have moved the backward operation in the main loop, thus the flag is no longer used.

**Comment by MysteryVaibhav** [comment]
@danishpruthi I have updated the code after your suggestions. Please review it again whenever you are free.

File: 03-wordemb-pytorch/wordemb-skip.py [hunk 5]
```diff
@@ -3,29 +3,23 @@
...
-        super(Skip, self).__init__()
+        super(WordEmbSkip, self).__init__()
 
         """ word embeddings """
         self.word_embedding = torch.nn.Embedding(nwords, emb_size)
         # uniform initialization
         torch.nn.init.uniform_(self.word_embedding.weight, -0.25, 0.25)
         """ context embeddings"""
-        self.context_embedding = torch.nn.Embedding(nwords, emb_size)
-        # uniform initialization
```
Comment (danishpruthi) [inline 03-wordemb-pytorch/wordemb-skip.py]: Is there any particular reason why you didn't initialize the context embeddings?

**Comment by danishpruthi** [comment]
@MysteryVaibhav: this looks good to me. Thanks.

@neubig please merge this.
